### PR TITLE
OCLOMRS-532: When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary

### PR DESCRIPTION
**Pull request by myrdstom**
# JIRA TICKET NAME:
[When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary](https://issues.openmrs.org/browse/OCLOMRS-532)

# Summary:
When adding a Q and A CIEL concept, the answer concepts should also be added to the dictionary

### CODE REVIEW

File: src/tests/Dashboard/action/api.test.js [hunk 1]
```diff
@@ -6,7 +6,12 @@ import { FETCHING_ORGANIZATIONS, ADDING_DICTIONARY } from '../../../redux/action
 import {
   fetchingOrganizations, createDictionary, createDictionaryUser,
 } from '../../../redux/actions/dictionaries/dictionaryActionCreators';
+// import {
```
Comment (Karuhanga) [inline src/tests/Dashboard/action/api.test.js]: Could we do away with these comments?

Comment (myrdstom) [reply src/tests/Dashboard/action/api.test.js]: Noted, on it

File: src/redux/actions/concepts/addBulkConcepts/index.js [hunk 2]
```diff
@@ -57,15 +60,37 @@ export const previewConcept = id => (dispatch, getState) => {
   return dispatch(isSuccess(payload[0], PREVIEW_CONCEPT));
 };
 
-export const addConcept = (params, data, conceptName) => async (dispatch) => {
+
+export const recursivelyFetchConceptMappings = async (fromConceptCodes, levelsToCheck) => {
+  const startingConceptMappings = await api.mappings.fetchFromPublicSources(fromConceptCodes.join(','));
+  const mappingsList = [startingConceptMappings.data];
+  for (let i = 0; i < levelsToCheck; i += 1) {
```
Comment (dkayiwa) [inline src/redux/actions/concepts/addBulkConcepts/index.js]: Is levelsToCheck different from LEVELS_TO_CHECK?

File: src/tests/Dashboard/action/api.test.js [hunk 3]
```diff
@@ -211,3 +212,25 @@ describe('Test suite for dictionary actions', () => {
...
+  beforeEach(() => {
+    moxios.install(instance);
+  });
+
+  afterEach(() => {
+    moxios.uninstall(instance);
+  });
+
+  it('should call the mappings endpoint with the right parameters', async () => {
+    const fromConcepts = '1,2,3';
```
Comment (dkayiwa) [inline src/tests/Dashboard/action/api.test.js]: Where are you asserting from?

File: src/tests/bulkConcepts/actions/bulkConcept.test.js [hunk 4]
```diff
@@ -233,6 +237,65 @@ describe('Test suite for addBulkConcepts async actions', () => {
...
+  let fetchFromPublicSourcesMock;
+
+  beforeEach(() => {
+    fetchFromPublicSourcesMock = jest.fn()
+      .mockReturnValueOnce({ data: [mappings[0]] })
+      .mockReturnValueOnce({ data: [mappings[1]] });
+    api.mappings.fetchFromPublicSources = fetchFromPublicSourcesMock;
+  });
+
+  it('returns a single level of mappings when called with 0 levelsToCheck', async () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/actions/bulkConcept.test.js]: The description does not follow the convention of the rest.

File: src/redux/actions/concepts/addBulkConcepts/index.js (hunk 2 above)
Comment (myrdstom) [reply src/redux/actions/concepts/addBulkConcepts/index.js]: They are the same. I thought it better to equate it to a constant as opposed to leaving it as a figure. Given the naming conventions used when assigning previous constants, I thought it better to maintain the standard

Comment (dkayiwa) [reply src/redux/actions/concepts/addBulkConcepts/index.js]: Have your team mates seriously reviewed all this?

File: src/tests/Dashboard/action/api.test.js (hunk 3 above)
Comment (myrdstom) [reply src/tests/Dashboard/action/api.test.js]: I have adjusted this

File: src/tests/bulkConcepts/actions/bulkConcept.test.js (hunk 4 above)
Comment (myrdstom) [reply src/tests/bulkConcepts/actions/bulkConcept.test.js]: I have adjusted the description to follow conventions

File: src/redux/actions/concepts/addBulkConcepts/index.js (hunk 2 above)
Comment (myrdstom) [reply src/redux/actions/concepts/addBulkConcepts/index.js]: I have shared this with the team and we need more clarity. `levelsToCheck` is a parameter, `LEVELS_TO_CHECK` represents the value that param takes. `LEVELS_TO_CHECK` currently is equal to the value `2`. Is the naming of the constant the issue or could you shed more light on what exactly we are missing?

File: src/components/dictionaryConcepts/components/helperFunction.js [hunk 5]
```diff
@@ -79,5 +79,5 @@ export const ATTRIBUTE_NAME_SOURCE = 'source';
 export const KEY_CODE_FOR_ENTER = 13;
 export const KEY_CODE_FOR_ESCAPE = 27;
 export const KEY_CODE_FOR_SPACE = 32;
-
+export const LEVELS_TO_CHECK = 2;
```
Comment (dkayiwa) [inline src/components/dictionaryConcepts/components/helperFunction.js]: When i look in this file, what does LEVELS_TO_CHECK mean or stand for?

Comment (myrdstom) [reply src/components/dictionaryConcepts/components/helperFunction.js]: `LEVELS_TO_CHECK`is a constant that determines how deep the `recursivelyFetchConceptMappings` function should go when looking for mappings to a parent Concept. I have renamed it to `MAPPINGS_RECURSION_DEPTH`

File: src/tests/bulkConcepts/actions/bulkConcept.test.js [hunk 6]
```diff
@@ -233,6 +237,65 @@ describe('Test suite for addBulkConcepts async actions', () => {
...
+
+  it('should return a single level of mappings when called with 0 levelsToCheck', async () => {
+    fetchFromPublicSourcesMock.mockClear();
+    const result = await recursivelyFetchConceptMappings([conceptCode], 0);
+    expect(fetchFromPublicSourcesMock).toHaveBeenCalledTimes(1);
+    expect(fetchFromPublicSourcesMock).toHaveBeenCalledWith(conceptCode.toString());
+    expect(result).toEqual([mappings[0].to_concept_url]);
+  });
+
+  it('should return two levels of mappings when called with 1 levelsToCheck', async () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/actions/bulkConcept.test.js]: Why doesn't it return the same levels of mappings as the levelsToCheck?

File: src/tests/bulkConcepts/actions/index.test.js [hunk 7]
```diff
@@ -15,18 +16,39 @@ import
...
+  api.mappings.fetchFromPublicSources = jest.fn(() => ({ data: [] }));
+
   beforeEach(() => {
     moxios.install(instance);
   });
 
   afterEach(() => {
     moxios.uninstall(instance);
   });
+  it('dispatches an error on fetchConceptSources failure', () => {
```
Comment (dkayiwa) [inline src/tests/bulkConcepts/actions/index.test.js]: Is this following our convention?

Comment (myrdstom) [reply src/tests/bulkConcepts/actions/index.test.js]: I have adjusted the description to better fit conventions. It is worth noting that all tests in this file follow the previous convention i.e something similar to `it(dispatches an error on fetchConceptSources failure)`. If the change is satisfactory, I will create a ticket to adjust the test descriptions in this file

File: src/tests/bulkConcepts/actions/bulkConcept.test.js (hunk 6 above)
Comment (myrdstom) [reply src/tests/bulkConcepts/actions/bulkConcept.test.js]: `levelsToCheck` is the second level of mappings i.e If a concept has mappings, it checks those mappings for mappings and returns those mappings too. When ` 1 levelToCheck` happens, it then returns the mappings of a concept and the mappings of the mappings of the same concept

**Pull request merged**